### PR TITLE
With custom HTML Renderer

### PR DESCRIPTION
**Pull request by kristojorg**
This is a draft PR showing what it would look like to remove R2D2BC.

TODO:

- [x] Color mode
- [x] Font size
- [x] Font family
- [x] Paginated mode
- [x] Snap to page in paginated mode
- [x] TOC navigation
- [x] Internal navigation links
- [x] Make internal link clicks register a change of location so prev/next page work
- [x] fix scroll bars
- [x] Remember where user is when switching between paginated and scroll mode
- [x] Scroll to end of resource when navigating backwards
- [x] Keep location in URL bar
- [x] paginated window resize
- [x] Fix bug when there is a "half page"

### CODE REVIEW

**Comment by kristojorg** [comment]
Pagination isn't working here: https://web-reader-git-refactor-remove-r2d2bc-nypl.vercel.app/v2/moby-epub2?location=%7B%22href%22%3A%22https%3A%2F%2Fweb-reader-git-refactor-remove-r2d2bc-nypl.vercel.app%2Fsamples%2Fmoby-epub2-exploded%2FOEBPS%2F%40public%40vhost%40g%40gutenberg%40html%40files%402701%402701-h%402701-h-1.htm.html%22%2C%22type%22%3A%22application%2Fxhtml%2Bxml%22%2C%22locations%22%3A%7B%22progression%22%3A0.7692144892144892%2C%22position%22%3A11%2C%22remainingPositions%22%3A2%7D%7D

**Comment by kristojorg** [comment]
@Toxiapo @hokei @samanthaandrews this PR is now ready for review. There is an architecture document at /HtmlReader/architecture.md. Take a look and please let me know your questions so I can update the arch document with answers!

**Comment by kristojorg** [comment]
I think once we have reviewed this, we should come up with a list of edge cases to add to our test suite so we can make sure this is running 100%

Top of my head:

- Fixed layout
- Initial location
- Reads location from query param on initial load. Handles multiple types of locators.
- Handles incoherent query param
- Internal link clicks (and location and query param update)
- Renders iframe title
- Forward/back buttons on browser work as expected (what is expected)
- hash navigation

File: src/HtmlReader/architecture.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+Readium-CSS also enables a "paginated" mode, where the full HTML resource is broken into multiple columns that are then rendered side by side in a horizontally scrolling container. Thus, to move to the next page in paginated mode, we simply scroll the user horizontally by one page-width.
+
+## Lifecycle and State
+
+We use `React.useReducer` as the single source of truth for state in this hook. Whenever anything changes, it should go through `useReducer`. If you render with the same state object twice, you should always get the same thing.
+
+Given that there are more than 10 properties in the state object, and each one has multiple possible values, there are seemingly infinite possible states. However, the range of _valid_ states is actually much smaller. These states are:
+
+- `InactiveState`: The hook was not provided any arguments, and thus it is inactive.
```
Comment (hokei) [inline src/HtmlReader/architecture.md:17]: Context:  So I'm starting my review/exploration with this file.  

Is it convention to call everything ThingState rather than ActiveState.FetchingResource (eg: https://github.com/NYPL-Simplified/web-reader/pull/91/files#diff-40c72fa84cd3e89b4e8825c1858a53ddb9d56a42805fcf97192b24aabbefe068R20)?

If someone were to say "oh my state is special, it's gotta be a new state" - where would they add it? How would they determine if that's actually true or not?

File: src/HtmlReader/architecture.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+- `NavigatingState`: The iframe has finished loading, and now we are waiting for our `useEffect` hook to complete any inter-resource navigation that might be necessary depending on the current `state.location`. For example, scrolling the user to a specific page.
+- `ReadyState`: We have now navigated the user to a specific page, and are now finished with our work.
+
+The definitions for these states can be found in `./types.ts`. Note that state transitions should happen in the order listed above, with the exception that it is possible to go backwards occasionally. For example, from the `ReadyState`, we may go back to `FetchingResourceState` if the user switches to a new resource, or we might simply go back to the `NavigatingState` if they navigate to a new location in the same resource.
+
+We provide the text value `state.state` as a string describing the state. This makes it very easy to perform type narrowing. If `state.state === 'READY'`, typescript knows that `state.iframe` and `state.resource`, and more all exist.
```
Comment (hokei) [inline src/HtmlReader/architecture.md:27]: should those values stay strings or should we enum them

File: src/HtmlReader/architecture.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+## Fetching resources
+
+When the state is `FetchingResourceState`, we need to actually go out and fetch the resource referenced in `state.location.href`. We do this in the `useResource` hook. When we transition to a `FetchingResourceState`, that hook will call `getContent` to get the resource. It then will dispatch a `RESOURCE_FETCH_SUCCESS` action, which will transition the state to `RenderingIframeState`.
+
+### Updating the resource
+
+After fetching the resource, there are a couple things we need to do to it before we can render it:
+
+- Set a `<base>` element in the `<head>` with the url of the resource. This tells the iframe where relative urls within the resource are relative to.
+- Inject any injectables. For example, we need to add the Readium-CSS stylesheets to the `<head>`.
```
Comment (hokei) [inline src/HtmlReader/architecture.md:42]: Is CSS the only injectible?  What other injectibles should we consider? Do these have implications for the rest of the update-resource process?

File: src/HtmlReader/architecture.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+- Set initial CSS variables like scroll settings and color mode.
+- Add a simple JS script that listens for internal link clicks and calls `window.parent.postMessage` to send a message to the parent about the clicked link.
+
+Note that we _could_ do all of this in a `useEffect` hook after the resource loads, but then there would be a flash of unstyled content between the first render and when we add our Readium-CSS stylesheet. To avoid this, we simple add it to the document string before it ever gets to the `<iframe>` for rendering.
```
Comment (hokei) [inline src/HtmlReader/architecture.md:46]: > we simple add it to the document string

nit: typo, but also:  what does the "it" here refer to?  The Readium-css stylesheet?

File: src/HtmlReader/architecture.md (hunk 2 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:27]: Enums aren't very nice in TS unfortunately, and strings provide all the same benefits. They are _typed_ strings though, so you can't accidentally put something different in there

File: src/HtmlReader/architecture.md (hunk 1 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:17]: That's a good question, I should think about this a little more. You're asking whether these are all a "sub-state" of the ActiveState?

In terms of how to know when you need a new one... It's a little bit arbitrary. When I first built this, I didn't have these split into multiple states. It was all just one `ReaderState`, but then I constantly had to check `if(state.isIframeLoaded && state.iframe && state.resource)` etc. So to minimize the number of those checks, I separated them into distinct states. Then we know that `if(state.state === 'READY')` then all those things are defined. It's just a convenient way to separate the complex number of options into distinct states so we can know what is defined at different times.

File: src/HtmlReader/architecture.md (hunk 3 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:42]: No, another injectable is JS or, theoretically, any element that can go in the head of the document. Typically, we should be prepared for CSS and JS injections. They don't have implications because we don't support _changing_ the injectables after initialization. We don't need that, and I think if someone wanted it, we would let them do it with some kind of API. But for now, no, the injectables are injected only when the resource is fetched.

File: src/HtmlReader/architecture.md (hunk 4 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:46]: Yea, I meant "we simply add all injectables to the document before it gets to the iframe".

Comment (hokei) [reply src/HtmlReader/architecture.md:46]: OK so it goes document -> put injectibles into document -> document + injectibles get to the iFrame.  And that's we make sure things are styled on first render.  Got it.

File: src/HtmlReader/architecture.md (hunk 1 above)
Comment (hokei) [reply src/HtmlReader/architecture.md:17]: > You're asking whether these are all a "sub-state" of the ActiveState?

Yes, or if they're all a "sub-state" of _something_.  clearly InactiveState isn't a substate of ActiveState.  The idea is to try to put all this together so that it's referenced everywhere, so that, for example, we don't have a chance of doing something like `state.state === 'NOT_QUITE_READY'` in a random place in the code.  I haven't looked through everything in detail so you might have put guardrails against that already.

Comment (hokei) [reply src/HtmlReader/architecture.md:17]: In essence, understanding all the states the reader can be in is the most important thing, so I'm wondering if there's a way we can define all of it in one place in the code, as well as in the documentation.  ActiveState seemed like the most obvious place to do it.

File: src/HtmlReader/lib.ts [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+ * Constants
+ */
+export const FONT_SIZE_STEP = 4;
+
+/**
+ * If we provide injectables that are not found, the app won't load at all.
+ * Therefore we will not provide any default injectables.
+ * @todo - this is not true in our custom renderer. We should provide default injectables.
+ */
+export const defaultInjectables: Injectable[] = [];
```
Comment (hokei) [inline src/HtmlReader/lib.ts:17]: what is the difference between fixed injectibles and non-fixed injectibles?

File: src/HtmlReader/index.tsx [hunk 6]
```diff
@@ -1,396 +1,262 @@
...
   ReaderReturn,
   ReaderArguments,
   FontFamily,
+  HtmlNavigator,
 } from '../types';
-import HtmlReaderContent from './HtmlReaderContent';
-import { Locator } from '@d-i-t-a/reader';
-import '../../node_modules/@d-i-t-a/reader/dist/reader.css';
+import LoadingSkeleton from '../ui/LoadingSkeleton';
+import { DEFAULT_HEIGHT, DEFAULT_SHOULD_GROW_WHEN_SCROLLING } from '..';
```
Comment (hokei) [inline src/HtmlReader/index.tsx]: where are these importing from?  I don't think I've ever seen this before haha

File: src/HtmlReader/index.tsx [hunk 7]
```diff
@@ -1,396 +1,262 @@
...
- * If we provide injectables that are not found, the app won't load at all.
- * Therefore we will not provide any default injectables.
+ * @TODO :
+ *
+ * - WORKING ON
+ *  - provide default injectables (Readium CSS)
+ *  - fix navigate to hash.
+ *
+ * Future:
+ *  - Don't use ReadiumCSS for fixed layout
```
Comment (hokei) [inline src/HtmlReader/index.tsx:35]: Why?

File: src/HtmlReader/index.tsx [hunk 8]
```diff
@@ -1,396 +1,262 @@
...
 /**
- * If we provide injectables that are not found, the app won't load at all.
- * Therefore we will not provide any default injectables.
+ * @TODO :
+ *
+ * - WORKING ON
+ *  - provide default injectables (Readium CSS)
+ *  - fix navigate to hash.
+ *
+ * Future:
```
Comment (hokei) [inline src/HtmlReader/index.tsx:33]: Are these all things that work in the previous reader?  If not, worth thinking about extracting them and letting POs prioritize

File: src/HtmlReader/index.tsx (hunk 7 above)
Comment (kristojorg) [reply src/HtmlReader/index.tsx:35]: It doesn't support fixed layout. Instead you rely on the publisher's CSS

File: src/HtmlReader/index.tsx (hunk 6 above)
Comment (kristojorg) [reply src/HtmlReader/index.tsx]: Should be the index two levels up I guess? I'm not sure why they auto imported from there. They are also defined in constants

File: src/HtmlReader/lib.ts (hunk 5 above)
Comment (kristojorg) [reply src/HtmlReader/lib.ts:17]: One is for fixed layout and the other is for reflowable books

File: src/HtmlReader/architecture.md (hunk 4 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:46]: Exactly

File: src/HtmlReader/architecture.md (hunk 1 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:17]: I suppose that's what I was trying to do in the `/HtmlReader/types.ts` file where I defined all the state types. Do you think that works?

File: src/HtmlReader/index.tsx (hunk 8 above)
Comment (kristojorg) [reply src/HtmlReader/index.tsx:33]: No, I'm pretty sure this PR is now feature-paired with prod. We should probably do some fixed layout testing though come to think of it. I need to update these comments

File: src/HtmlReader/architecture.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+- `progression`: A decimal value representing percent of the way through the resource. This is what we typically use when navigating to scroll positions. We calculate the `progression` of the desired position and scroll the user there.
+- `fragment`: A string representing a location. Currently, we only support anchors (eg: `"#some-heading"`). We can scroll the user there by querying the dom for that element and then calling `el.scrollIntoView()`.
+- `position`: An integer representation of the location, which I interpret to mean "page number". This is not very portable because the number of pages and size of pages changes when browser size changes or font size changes. Thus, we prefer to use `progression`, but we keep `position` updated in case we wish to display it at some point.
+
+Note on CFIs: In the future, we would like to support CFIs. These are strings representing specific locations in an HTML document. They are portable as well. When you're reading a kindle and it says "location 441/2387", those numbers are the CFI that you are on and the total number of CFIs. More work on this is needed in the future.
```
Comment (Toxiapo) [inline src/HtmlReader/architecture.md:63]: What does CFI stand for?

File: src/HtmlReader/architecture.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+1. `useLocationQuery`: We now keep a copy of `state.location` sync'd to the `?location` query param in the URL. This allws users to share or save links to specific locations. So we use an effect to update the query whenever `state.location` changes.
+1. `useWindowResize`: When the window resizes, we need to re-scroll the user to the location they are meant to be on. Thus we need to move backwards to the `NavigatingState`.
+1. `useIframeLinkClick`: Uses `React.useEffect` to add a event listener for a "message" event, which is sent from the iframe upon internal link clicks. We then dispatch a `GO_TO_HREF` with the clicked value.
+
+## Decisions
+
+- Store all state in `useReducer`.
+- Use the `webpubManifestUrl` as the base url when constructing urls from relative urls. If we get a webpub manifest with relative urls, we must assume they are relative to the webpub manifest url itself. We always resolve urls to full URL objects instead of leaving them as possible relative strings.
```
Comment (Toxiapo) [inline src/HtmlReader/architecture.md:75]: This is great that we made this explicit! This makes things a lot easier (i.e. when comparing internal links. 😆 )

File: src/HtmlReader/effects.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+  );
+  setCSSProperty(
+    html,
+    '--USER__fontFamily',
+    familyToReadiumFamily[state.fontFamily]
+  );
+  setCSSProperty(html, '--USER__fontSize', `${state.fontSize}%`);
+  setCSSProperty(html, 'overflow', state.isScrolling ? 'scroll' : 'hidden');
+  // set the number of columns to only ever have 1.
+  setCSSProperty(html, '--USER__colCount', '1');
```
Comment (Toxiapo) [inline src/HtmlReader/effects.ts:99]: 1 column FTW 🎉

File: src/HtmlReader/lib.ts [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+  link: ReadiumLink,
+  baseUrl: string,
+  locations: Locator['locations'] = {}
+): Locator {
+  const url = new URL(link.href, baseUrl);
+  const hash = url.hash;
+  const hrefWithoutHash = url.origin + url.pathname;
+  // add the hash if we don't already have a fragment
+  if (hash && !locations?.fragment) locations.fragment = hash;
+  // const hrefWithoutHash =
```
Comment (Toxiapo) [inline src/HtmlReader/lib.ts]: nit: dangling comment

File: src/HtmlReader/architecture.md (hunk 9 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:63]: Canonical Fragment Identifier. I'm not totally sure how they work, but basically it's some string that points to a specific position in an HTML document.

File: src/HtmlReader/architecture.md (hunk 10 above)
Comment (kristojorg) [reply src/HtmlReader/architecture.md:75]: Yea that was a pain previously!

File: src/HtmlReader/lib.ts [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+  script.setAttribute('type', 'text/javascript');
+  script.textContent = `
+    window.onload = () => {
+      window.top.postMessage({source: 'reader-iframe', type: 'IFRAME_LOADED'})
+    }
+  `;
+  document.head.appendChild(script);
+}
+
+export function getMaybeIframeHtml(
```
Comment (hokei) [inline src/HtmlReader/lib.ts:232]: dumb question - there truly isn't a better way to do an optional?

Comment (hokei) [reply src/HtmlReader/lib.ts:232]: Is this so that you don't have to check for `undefined` every time you use `getIframeHtml`?

**Review by hokei — APPROVED** [review]
This looks really good.  

Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?  

It can be manual tests.  

(This can be a new ticket) 
Do we have automated test suites for all of our locations logic?  I see that navigations has page button stuff, but (it might be there but I might be missing it?) something that tests things like, when you scroll to the bottom of the chapter and press "next", it knows that it's at the end of the chapter and loads the next one.  Or, when you scroll, then resize, it knows you're not yet at the end of the chapter.  

(Or are these things no longer edge cases?) 

Also, do we need a tests for the example app, for things like making sure that embedded mode continues to work?

**Comment by hokei** [comment]
(not) fun scrollbar bug, I think
Screen Shot 2022-01-24 at 1 45 17 PM

  (this is chrome with a mouse plugged in - you can't see it in regular view because it's white) 
  
 (This is the kind of thing that I don't really know how to test for, but also is the kind of thing I worry about regressions on)

File: src/HtmlReader/useResource.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        // set the initial CSS state
+        setCss(document.documentElement, {
+          colorMode: state.colorMode,
+          fontSize: state.fontSize,
+          fontFamily: state.fontFamily,
+          isScrolling: state.isScrolling,
+        });
+
+        // inject js to communicate with iframe
+        // injectJS(document);
```
Comment (samanthaandrews) [inline src/HtmlReader/useResource.ts]: nit: remove unused injectJS() since we are calling it above

**Comment by Toxiapo** [comment]
> This looks really good.
> 
> Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?
> 
> It can be manual tests.
> 
> (This can be a new ticket) Do we have automated test suites for all of our locations logic? I see that navigations has page button stuff, but (it might be there but I might be missing it?) something that tests things like, when you scroll to the bottom of the chapter and press "next", it knows that it's at the end of the chapter and loads the next one. Or, when you scroll, then resize, it knows you're not yet at the end of the chapter.
> 
> (Or are these things no longer edge cases?)
> 
> Also, do we need a tests for the example app, for things like making sure that embedded mode continues to work?

Yeah, I agree! I think some manual tests are good for this PR and getting more tests coverages in the future.

(Thinking out loud on some of the issues we had faced previously)

1. Internal links works on both paginated and scrolling mode
2. TOC navigates on paginated and scrolling mode
3. Height and width problem, make sure there are no extra scrollbars 
4. All the settings under the settings button
5. Next / Previous button response to book state (navigates + hide/show state), i.e stuck on cover image issue
6. Switch from paginated to scrolling mode saves the reading position, vice versa

File: src/HtmlReader/lib.ts (hunk 13 above)
Comment (kristojorg) [reply src/HtmlReader/lib.ts:232]: Exactly. I wanted one function for when it _has_ to be defined, and another for when maybe it's defined and maybe not.

**Comment by kristojorg** [comment]
> Do we have a good set of acceptance criteria on how to test this, to make sure that it truly is feature-parity with the previous reader?

I don't have anything concrete, but Jiayong's list is a good starting point. I would just add fixed layout to that. Is this something I should do or I should have the QA team do?

I updated [OE-321](https://jira.nypl.org/browse/OE-321) to add automated tests for these edge cases you listed above. I think that will be very helpful. Probably won't happen for me until next sprint though.

**Comment by kristojorg** [comment]
@hokei RE that scroll bar bug:

Soooo I don't think this is a regression. Question: Can you actually scroll horizontally? How much if so? The issue is that when scroll bars are showing, which you see when you have a mouse plugged in (can also enable it in system preferences/general), the vertical scroll bar actually makes the available space to render smaller, but is not included in `window.innerWidth` or `clientWidth`. So when we calculate the width of the window to render, we might be off by a certain amount due to the vertical scrollbar. Which causes a horizontal scrollbar to be rendered (lol). We _could_ try to figure out if it is showing, but then there is the secondary issue that it is actually a different size on each platform/browser combo. Sooooo I just decided it's not worth it basically. If you're using the app with scrollbars showing like that, you're probably used to seeing scrollbars everywhere even when there is very little area to scroll. It seems like a platform problem to me

**Comment by Toxiapo** [comment]
I am seeing a UI bug with the internal link, this does not affect the user's ability to read the book, b/c it's fine if you refresh the page or change the page using next/prev button, so maybe we can fix this after the merge? Unless this is a quick fix.

To reproduce:
1. Go on /html/moby-epub2
2. Go to the page with internal links, and click on one
3. The page navigates but the page layout is misaligned.  

Screen Shot 2022-01-26 at 10 02 50 AM

**Comment by hokei** [comment]
>  Question: Can you actually scroll horizontally? How much if so? 

Screen Shot 2022-01-26 at 10 11 28 AM

So yeah it's a bit of a strange bug in that I see the scrollbars area but I cannot scroll.  I've showed it in the embedded view which makes this super clear that scrolling isn't actually possible. I also think it isn't a high priority bug and should not block merge.  Honestly I think very few people will notice it and we can fix it when they start filing bug reports (or when I get annoyed enough with it haha)

**Comment by hokei** [comment]
I tried putting this in next apps (CPW and DRB) and got this error: 
Screen Shot 2022-01-26 at 3 46 37 PM

Does there need to be updates to the Readme on how to incorporate the CSS?

File: src/HtmlReader/index.tsx (hunk 6 above)
Comment (kristojorg) [reply src/HtmlReader/index.tsx]: Fixed this

File: cypress/integration/html-reader/navigations.ts [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
-    cy.loadPage('/streamed-alice-epub');
+    cy.loadPage('/html/streamed-alice-epub');
   });
 
   it('should contain a link to return to the homepage', () => {
     cy.findByRole('link', { name: 'Return to Homepage' }).should(
       'have.prop',
       'href',
-      `${Cypress.config().baseUrl}/`
+      `${Cypress.config().baseUrl?.replace('/v1', '').replace('/v2', '')}/`
```
Comment (Toxiapo) [inline cypress/integration/html-reader/navigations.ts]: Do we still need to do this?

File: .github/workflows/cypress.yml [hunk 16]
```diff
@@ -27,7 +33,8 @@ jobs:
           key: my-cache-${{ runner.os }}-${{ hashFiles('package-lock.json') }}
 
       - name: Install Cypress 📥
-        run: npm i cypress
+        # don't install every node_module however
+        run: npm install --no-package-lock --no-save cypress
```
Comment (Toxiapo) [inline .github/workflows/cypress.yml]: I am curious what this does? How does the CI install other packages?

File: cypress/integration/html-reader/errors.ts [hunk 17]
```diff
@@ -17,25 +17,23 @@ describe('error states', () => {
...
+    cy.visit('/html/test/missing-resource');
 
     cy.getIframeBody().contains('Page not found.', { timeout: 10000 });
   });
 
   it('throws error for missing injectable', () => {
-    cy.visit('/test/missing-injectable');
+    cy.visit('/html/test/missing-injectable');
 
     cy.findByRole('heading', { name: 'An error occurred' });
```
Comment (Toxiapo) [inline cypress/integration/html-reader/errors.ts]: We probably don't need this since we will always have default injectables right?

File: src/Readium/Locator.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
+export interface Locator {
+  href: string;
+  type?: string;
+  title?: string;
+  locations: Locations;
+  text?: LocatorText;
+  // displayInfo?: any;
```
Comment (samanthaandrews) [inline src/Readium/Locator.ts]: nit: remove displayInfo

**Comment by hokei** [comment]
Has anyone been able to `npm link` this into a consuming package such as CPW?  I just want to make sure that this has been tested - I think I'm running into react version errors, which might be  npm link errors rather than errors in the new code, so I don't want to that to hold up merge -- just let me know

File: cypress/integration/html-reader/navigations.ts (hunk 15 above)
Comment (kristojorg) [reply cypress/integration/html-reader/navigations.ts]: Good point

**Comment by kristojorg** [comment]
>Has anyone been able to npm link this into a consuming package such as CPW? I just want to make sure that this has been tested - I think I'm running into react version errors, which might be npm link errors rather than errors in the new code, so I don't want to that to hold up merge -- just let me know

No I haven't either. It's notoriously difficult to npm link with react. Involves a lot of BS. I am planning to publish an alpha and test with that via npm in CPW

**Comment by hokei** [comment]
> Involves a lot of BS. I am planning to publish an alpha and test with that via npm in CPW

OK that sounds like a good plan! Let me know and I'll test it on DRB too.

File: src/types.ts [hunk 19]
```diff
@@ -1,4 +1,5 @@
 import { Injectable } from '@d-i-t-a/reader/dist/types/navigator/IFrameNavigator';
```
Comment (Toxiapo) [inline src/types.ts]: old d-i-t-a reference

**Review by samanthaandrews — APPROVED** [review]
I have done a bunch of manual testing as well as reviewing the code and this looks really good. I know @Toxiapo and @hokei have more experience with this code base, and this PR has gone through thorough code review by them.

**Comment by kristojorg** [comment]
@hokei I have published as `@nypl/web-reader@1.1.0-0`. Install that version to test. I tested in CPW and all is working well so far.

**Comment by hokei** [comment]
This looks really good, and I think we should merge it.  However, I'll note down the things that I noticed.  I don't think we need to maintain two forks or anything - just that between OE and DRB we need to figure out when to fix these things before DRB can upgrade.  

Scrollbars:  left is current reader, right is new reader.  
Screen Shot 2022-03-07 at 9 33 14 AM

TOC: 
Screen Shot 2022-03-07 at 9 34 00 AM
All of the TOC items seem to have two borders instead of one ( this is a problem with everything, but it's most noticeable in the TOC) 

And then, 1 more that I think is an issue in both the old and new renderers, but you can see it in the new renderer in the screenshot above: the TOC should be above the prev/next buttons, but they're not. 


aaaaand that's really all I can find!  This is amazingly impressive work and let's get it in!

**Comment by hokei** [comment]
Also Kristo - after you merge, can you write up a little thing that's just what this feature will do for us?  Something that we can share with Miguel and Sean - what is now possible that was impossible before?  Especially for Sean, who is probably more likely to be making WR changes in the next 6-12 months

**Review by Toxiapo — APPROVED** [review]

**Review by hokei — APPROVED** [review]
